### ISSUE TITLE
Suggest case x of Just y -> z ; Nothing -> return () -> for_ x $ \y -> z

### ISSUE THREAD

**Issue by ocharles** [ndmitchell/hlint#424]
What do you think of blessing this rule?

```yaml
- warn: 
    lhs: case x of Just y -> z ; Nothing -> return ()
    rhs: for_ x $ \y -> z
```

**Comment by ndmitchell** [comment]
I am somewhat torn. The problem is that `for_` is misnamed - it should be `forM_` since it's just the generalisation of `Foldable` over `forM_`. It's also confusing in the sense that `for` sounds very "looping" like, when in this case it's really more of an `if` or `when`. If there was a [`whenJust`](https://hackage.haskell.org/package/extra-1.6.2/docs/Control-Monad-Extra.html#v:whenJust) in the base library I'd have definitely made this a rule already, but pushing people towards `for_` seems a little unpleasant - although maybe that's just my grumpy "stay off my Prelude" hat on.

If we do add it, should probably also cover the cases where the branches are reverse, the `Nothing` is `_` and it's written with a `maybe` combinator.

**Comment by ocharles** [comment]
*shrug* `for_` is what we have, so that's just what we use - I never really gave it much thought.

> If we do add it, should probably also cover the cases where the branches are reverse, the Nothing is _ and it's written with a maybe combinator.

I'm not following this, what rule are you suggesting?

**Comment by ndmitchell** [comment]
I had already got `warn: {lhs: case m of Nothing -> return (); Just x -> f x, rhs: Data.Foldable.forM_ m f}`, so your variation is just that with the case expression the other way around! The use of `forM_` rather than `for_` makes me slightly happier, but of course they are equivalent.
